internal_ws/ykcompile/src/arch/x86_64/mod.rs

```diff
@@ -409,6 +409,7 @@ impl TraceCompiler {
         self.stack_builder.alloc(ty.size(), ty.align())
     }
 
+    /// Implements the StorageLive statement, assigning a Location the local.
```

Comment:
I'm not sure how to parse this sentence.

Reply:
How about:
> Called when a local variable becomes live. The register allocator is invoked to make storage space for the local.

Reply:
Which part is the problem? This basically tells the compiler a variable is live and assigns it a register or location on the stack. How about: `Implements a local going live, assigning it to a register or, if no registers are free, to a location on the stack.`

Reply:
@vext01 I think, in general, more active descriptions are clearer.

@lukasd Words like "implements" feel a bit like padding. I'm also not sure what "a local going live" exactly means.

Taking those thoughts together, would something like "Assign a Location to a StorageLive (i.e. finding it a register or space on the stack)." be accurate? Maybe too terse?

Reply:
`StorageLive` is a MIR statement, so saying we assign a location to it, doesn't sound right. We are assigning a location to the local that MIR has decided is live from that point onwards until we see a `StorageDead` for it.

Reply:
Tweaking Laurie's prose a bit:
> Assign a `Location` to a `Local` just turning live (i.e. if possible, find it a register to live in, or failing that, space on the stack).

Reply:
Somewhat related, but https://github.com/rust-lang/rust/pull/78360 if merged will remove `Storage*` after borrowck when not optimizing.

Reply:
@ptersilie does commenting out the call to `local_to_location` cause any test to fail?

If my mental model is correct, any later bytecode mentioning a local will already call `local_to_location` and thus lazily allocate space for it as-needed?

Reply:
Ah, it might bust stopgapping...

Reply:
Yeah, I was thinking the same, but seem to remember there was a slight difference, but will have to double check.

Reply:
Nope, looks like all tests pass, so this is just here to have the counter to `storagedead`. Though it may mean that a local can snatch a register if the live happens before data stored in it.

Reply:
> Somewhat related, but [rust-lang/rust#78360](https://github.com/rust-lang/rust/pull/78360) if merged will remove `Storage*` after borrowck when not optimizing.

This would be annoying for us (at least until we have our own liveness analysis), so I hope it can be re-enabled in ykrustc.

Reply:
> Nope, looks like all tests pass, so this is just here to have the counter to storagedead.

Does this mean we can delete / simplify code or ... ?

Reply:
I think we can remove it, though it is a bit odd that `StorageLive` then has no real purpose. Maybe we can use this function as a check that we don't `StorageLive` the same variable twice, without `StorageDead`ing it first. I could imagine a future bug in MIR/SIR/TIR missing a `StorageDead`, or `VarRenamer` not renaming inlined functions properly, which we otherwise wouldn't notice unless it causes programs to behave weirdly.

Reply:
It means that we don't strictly have to allocate a local until its first use, and could in fact remove the function in question altogether.

It might result in better allocation results overall. Consider the following:

```
live($1)
if (...) {
   $1 = call foo(...)
   call blah($1)
} else {
  // don't use $1
}
```

If we trace the else branch, then right now, then the trace would allocate space for, but never use $1. In other words, it could potentially block another variable from using a register.

If we allocate on first use, we wouldn't allocate $1 at all (in this trace), leaving it free for other locals.

But, I don't know if MIR would generate code like this...

Another question I then have is, if we don't allocate eagerly, what is the purpose of `StorageLive` in our bytecode?

Reply:
Lukas and I raced, but agreed.

So perhaps `StorageDead` is useful for #243. But then again, we could just consider the first use of a variable as its birth?

Reply:
I have no idea so I'm relying on you two to work this out :)

Reply:
> ```
> live($1)
> if (...) {
>    $1 = call foo(...)
>    call blah($1)
> } else {
>   // don't use $1
> }
> ```

I don't know why MIR wouldn't put the `live($1)` in the same block where it's assigned and used, but you never know. Although a quick test on Rust Playground seems to confirm this suspicion (which again doesn't mean MIR always does this).

Reply:
Fixed in 88f0a69 and a7b7aee.